### PR TITLE
Cobertura hardcoded URI fix and Jacoco fix 

### PR DESCRIPTION
**Pull request by notaturkey**
This is an implementation that someone already has a pull request for [(link here)](https://github.com/jenkinsci/code-coverage-api-plugin/pull/188), but this solution is more memory and time efficient. On top of finding paths, this change handles jacoco tested files that are the same name but live in different packages. This is done by including the package path to the src file name when parsing the jacoco xml.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Review by uhafner — COMMENTED** [review]
Thanks for providing these new changes! I walked through your code in a first iteration but I must admit that it is hard to follow the changes since you are mixing different things in one pull request. Additionally, you did not provide a single new test that exposes the problem you are actually trying to fix. So I am a little bit reluctant on how to proceed. 

Since I am maintaining this plugin for a couple weeks only I have some more additional questions in order to understand the reasoning for this PR. I am currently refactoring and replacing most of the old code into the new `model` package. This is a step by step approach and I did not yet find the time to look at the source code resolution. From a quick look over those classes it seems that here actually a lot of things are not ideal from an architecture point of view. So I need to make sure that your changes do not make it even worse. Currently we do not have good integration tests (e.g. for the source code resolution) but this is something I am going to address at my testing course here at the university. It would be helpful though, if you can add additional tests to your PR as well.

So this brings me to some key questions, what exactly is the problem with the file name resolution? It the problem, that the absolute file names are wrong after parsing the reports? Or is the problem that we have the same Filename in different folders? (I see that scanning the workspace for all files is already present in the current code base but I actually do not understand why that should be required, in the warnings plugin I do not need such a brute force approach anymore). Maybe we should discuss the reasoning for this change a little bit deeper. Maybe a totally different approach will help. Since the source code rendering (and copying) works quite good in my other plugins (warnings plugin) I think we can reuse some of the concepts there.

File: src/main/java/io/jenkins/plugins/coverage/adapter/JacocoReportAdapter.java [hunk 1]
```diff
@@ -65,13 +67,81 @@ public String getDisplayName() {
...
+        private static final Pattern METHOD_ARGS_PATTERN = Pattern.compile("\\[*([TL][^;]*;)|([ZCBSIFJDV])");
+        
         public JacocoCoverageParser(final String reportName) {
             super(reportName);
         }
 
         @Override
         protected CoverageResult processElement(final Element current, final CoverageResult parentResult) {
-            CoverageResult result = super.processElement(current, parentResult);
+            CoverageResult result = null;
```
[the remark this answers is not in the document]
Comment (notaturkey) [reply src/main/java/io/jenkins/plugins/coverage/adapter/JacocoReportAdapter.java:79]: the switch statement is really the only thing that changed from the JavaCoverageParser, but i cant think of a neater way of just only changing specific cases

File: src/main/java/io/jenkins/plugins/coverage/adapter/JacocoReportAdapter.java [hunk 2]
```diff
@@ -110,5 +180,72 @@ protected CoverageResult processElement(final Element current, final CoverageRes
...
         }
+
+        private String replacePathOrDollarWithDots(final String name) {
+            if (StringUtils.isNotBlank(name)) {
+                return name.replaceAll("[\\\\/$]", ".");
+            }
+            return name;
+        }
+
+        private String buildMethodName(final String name, final String signature) {
```
[the remark this answers is not in the document]
Comment (notaturkey) [reply src/main/java/io/jenkins/plugins/coverage/adapter/JacocoReportAdapter.java:191]: these are all methods taken from the JavaCoverageParser.java, they were private in that file so i was following the same style i guess.

File: src/test/java/io/jenkins/plugins/coverage/model/CoverageNodeTest.java [hunk 3]
```diff
@@ -157,7 +157,7 @@ void shouldSplitPackages() {
     void shouldComputeDelta() {
         CoverageNode tree = CoverageNodeConverter.convert(readResult("jacoco-analysis-model.xml"));
 
-        String checkStyleParser = "CheckStyleParser.java";
+        String checkStyleParser = "edu/hm/hafner/analysis/parser/checkstyle/CheckStyleParser.java";
```
[the remark this answers is not in the document]
Comment (notaturkey) [reply src/test/java/io/jenkins/plugins/coverage/model/CoverageNodeTest.java:160]: this is definitely an unintended bug

**Comment by notaturkey** [comment]
In response to your key questions:

**what exactly is the problem with the file name resolution?** 
when ant does the jacoco test and makes the xml to parse the filename is stored in a tag. since its a jacoco result, its picked up by JacocoReportAdapter.java. The filename received from the xml tag isnt very descriptive though which causes some problems.. 

**Is it the problem, that the absolute file names are wrong after parsing the reports? Or is the problem that we have the same filename in different folders? ** 
imagine a java project being tested with this structure:
```
src/main/java/org/example/ui/iconsPhone/button.java
src/main/java/org/example/ui/icons/button.java

```

In the released version of this plugin, the jacoco adapter makes it so the coverageNode would only be storing the "button.java" part of the filepath in the coverage result, and would only have that to find the source file to copy over if the user wants the source code views of the test results. Therefore this fix finds source files correctly and tweaks how its parsing results coming from jacoco to give descriptive paths.

**(I see that scanning the workspace for all files is already present in the current code base but I actually do not understand why that should be required, in the warnings plugin I do not need such a brute force approach anymore).** 
from what i understand the only reason is for the interactive views where jenkins presents you with the painted source code. If the jacoco xml gave absolute paths we wouldnt need to search but thats not the case. Also if the workspace changes after a jacoco test, those files can potentially be no longer there.


**Maybe we should discuss the reasoning for this change a little bit deeper. Maybe a totally different approach will help. Since the source code rendering (and copying) works quite good in my other plugins (warnings plugin) I think we can reuse some of the concepts there.**
Theres probably a better way of doing this that i havent come up with. The overall process of how this change works is: 

1. the plugin starts parsing the xml from jacoco
2. the CoverageResult tree gets populated from tags in the xml
3. On a package tag when the tree is being created, the raw value now gets attached to the node which is already a path of the package 
4. On a filename tag, the node created to store the filename now looks at its parent for the package path, if it cant find a parent thats a package, it defaults to how the plugin was previously storing the filename


 

Can you provide a link to the warnings plugin and what you think i should be using out of it?

File: src/main/java/io/jenkins/plugins/coverage/adapter/JacocoReportAdapter.java [hunk 4]
```diff
@@ -65,13 +67,81 @@ public String getDisplayName() {
...
+
+                    if (parent != null){
+                        result = new CoverageResult(CoverageElement.get("File"), parentResult,
+                            parentResult.getPackageSourcePath() + '/' + getAttribute(current, "name", null));
+                        result.setRelativeSourcePath(parentResult.getPackageSourcePath() + '/' + getAttribute(current, "name", null));
+                    }
+                    else{
+                        result = new CoverageResult(CoverageElement.get("File"), parentResult,
+                            getAttribute(current, "name", null));
+                        result.setRelativeSourcePath(getAttribute(current, "name", null));
```
[the remark this answers is not in the document]
Comment (notaturkey) [reply src/main/java/io/jenkins/plugins/coverage/adapter/JacocoReportAdapter.java:110]: what file is that getAbsolutePath method in and how is it being set for the Jacoco adapter?

File: src/test/java/io/jenkins/plugins/coverage/model/CoverageNodeTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (notaturkey) [reply src/test/java/io/jenkins/plugins/coverage/model/CoverageNodeTest.java:160]: A solution I can see to this is when the xml is being parsed in the file section, e.g line 96 of JacocoReportAdapter.java, we can just attach getAttribute(current, "name", null) to an instance of the CoverageResult object. But if this bad from an architecture perspective then id need some guidance.

File: src/main/java/io/jenkins/plugins/coverage/targets/CoverageResult.java [hunk 5]
```diff
@@ -145,6 +147,14 @@ public void stripGroup() {
...
     }
 
+    public void setPackageSourcePath(String path){
+        this.packagePath = path;
+    }
+
+    public String getPackageSourcePath(){
+        return this.packagePath;
+    }
+
```
[the remark this answers is not in the document]
Comment (notaturkey) [reply src/main/java/io/jenkins/plugins/coverage/targets/CoverageResult.java:157]: Im a bit confused about this question, you can't get these values till the xml is being parsed

**Comment by uhafner** [comment]
Thanks for your detailed answer!

> 
> **what exactly is the problem with the file name resolution?** when ant does the jacoco test and makes the xml to parse the filename is stored in a tag. since its a jacoco result, its picked up by JacocoReportAdapter.java. The filename received from the xml tag isnt very descriptive though which causes some problems..

Can you add a new test here that reads such a report and has assertions for the expected properties of file names?
It is still unclear for me what you are meaning by "isn´t very descriptive".

> 
> **Is it the problem, that the absolute file names are wrong after parsing the reports? Or is the problem that we have the same filename in different folders?** imagine a java project being tested with this structure:
> 
> ```
> src/main/java/org/example/ui/iconsPhone/button.java
> src/main/java/org/example/ui/icons/button.java
> ```
> 
> In the released version of this plugin, the jacoco adapter makes it so the coverageNode would only be storing the "button.java" part of the filepath in the coverage result, and would only have that to find the source file to copy over if the user wants the source code views of the test results. Therefore this fix finds source files correctly and tweaks how its parsing results coming from jacoco to give descriptive paths.

Ok. This is what I expected. So there are two different problems with your approach (and the current algorithm):
- The coverage painting algorithm currently works in the following way: after the XML Jacoco file has been parsed all file nodes are processed and the file (without a relative or absolute path) is searched for in the workspace. If found (it already uses a full scan of the file system) then this file is converted to an HTML file (with the coverage information) and then copied to Jenkins build folder.  Here we have the first problem: all sources are copied in the flat folder `builds/number/coverage-sources`. I.e., if the file name is the same, then duplicate files still will be overwritten. So your idea simply helps to find the correct source, but the destination still is broken.
- The second thing is that we currently do not use the context of the tree right now to create the correct path. You are trying to fix that by storing that path in the file nodes but I don't think that this is actually necessary. 

We have the following structure right now:

```
[Module] -  [Package1] - [File1]
       \             \  - [File2]
        \    - [Package2] - [File3]
   
``` 

Each package has the correct relative path. Each file has the correct file name. Then we can dynamically create the absolute filename as file.getParent().getFileName() + file.getFileName(). There is no need to store absolute paths in the file node. I think if we would reuse this information then everything will work out of the box.

> 
> **(I see that scanning the workspace for all files is already present in the current code base but I actually do not understand why that should be required, in the warnings plugin I do not need such a brute force approach anymore).** from what i understand the only reason is for the interactive views where jenkins presents you with the painted source code. If the jacoco xml gave absolute paths we wouldnt need to search but thats not the case. Also if the workspace changes after a jacoco test, those files can potentially be no longer there.

No, this is wrong. The workspace is not available anymore right after the build. All files are copied to the controller to save them for later access. 

> 
> **Maybe we should discuss the reasoning for this change a little bit deeper. Maybe a totally different approach will help. Since the source code rendering (and copying) works quite good in my other plugins (warnings plugin) I think we can reuse some of the concepts there.** Theres probably a better way of doing this that i havent come up with. The overall process of how this change works is:
> 
> 1. the plugin starts parsing the xml from jacoco
> 2. the CoverageResult tree gets populated from tags in the xml
> 3. On a package tag when the tree is being created, the raw value now gets attached to the node which is already a path of the package
> 4. On a filename tag, the node created to store the filename now looks at its parent for the package path, if it cant find a parent thats a package, it defaults to how the plugin was previously storing the filename
> 
> Can you provide a link to the warnings plugin and what you think i should be using out of it? Also after giving a better description of what this is doing I will wait to see if this fits from an architecture perspective before i write any tests.

So basically the copy operation should be as simple as that (pseudo-code):
```
void copy(CoverageFileNode file, workspaceFolder, buildFolder) {
    String content = workspaceFolder.open(file.getPath()).read(); 
    paint(content, file);
    File.create(buildFolder, file.getPath(), content);
```

That is all, no need to scan the whole file system.

**Comment by uhafner** [comment]
Here is also an example with the correct paths: https://ci.jenkins.io/job/Plugins/job/code-coverage-api-plugin/job/master/lastBuild/coverage/#coverageTable


So it would be really helpful if you can show an example where it does not work.

**Comment by notaturkey** [comment]
Thanks for your time, I appreciate your patience and attention to detail.

yes of course!
> Can you add a new test here that reads such a report and has assertions for the expected properties of file names? It is still unclear for me what you are meaning by "isn´t very descriptive".

not necessarily, the filename is now unique since it has the package path attached to it.
> So there are two different problems with your approach (and the current algorithm):
> 
>     * The coverage painting algorithm currently works in the following way: after the XML Jacoco file has been parsed all file nodes are processed and the file (without a relative or absolute path) is searched for in the workspace. If found (it already uses a full scan of the file system) then this file is converted to an HTML file (with the coverage information) and then copied to Jenkins build folder.  Here we have the first problem: all sources are copied in the flat folder `builds/number/coverage-sources`. I.e., if the file name is the same, then duplicate files still will be overwritten. So your idea simply helps to find the correct source, but the destination still is broken.
> 

This was my initial thought but when the code is trying to find the file to move to the controller, its passed a map (filename -> painted object). If we could instead just pass a list of the filenodes then we could do that approach but then youd have the same naming conflict problem in the controller directory unless you did ```file.getParent().getFileName() + file.getFileName()``` there also. Itd be a way bigger commit than what im already doing so thats why i was hesitant. Another important thing to consider is that this logic is done to all plugins. I'm if itd play well with something like cobertura. 
> Then we can dynamically create the absolute filename as file.getParent().getFileName() + file.getFileName(). There is no need to store absolute paths in the file node. I think if we would reuse this information then everything will work out of the box.
> 

I'm still sure we need to search the workspace. The path needs to be found during runtime since the xml paths from jacoco are all relative. If the paths from the xml were absolute, itd have still have to search in the case of someone running a jenkins job that uses jacoco to test, then moves the source code within the workspace, and finally tries to use the plugin with xml files that have the wrong paths. 
> And the other thing that we need is a new method `getPath()` that returns for packages the path to the package and for files the path from the parent (i.e. the package) concatenated with the plain file name.

**Comment by uhafner** [comment]
> Another important thing to consider is that this logic is done to all plugins.  I'm if itd play well with something like cobertura.

The Cobertura plugin is obsolete and not required anymore. The functionality is now here in this Code Coverage API plugin.

> > Then we can dynamically create the absolute filename as file.getParent().getFileName() + file.getFileName(). There is no need to store absolute paths in the file node. I think if we would reuse this information then everything will work out of the box.
> 
> I'm still sure we need to search the workspace. The path needs to be found during runtime since the xml paths from jacoco are all relative. If the paths from the xml were absolute, itd have still have to search in the case of someone running a jenkins job that uses jacoco to test, then moves the source code within the workspace, and finally tries to use the plugin with xml files that have the wrong paths.

I don't think is necessary. I don't do it in the warnings plugin, and it should not be required here as well. 

I'll have a look at the code to see if I can refactor it in the following way:

- run the XML parser on the agent: 
   - parse the files into coverage results
   - for each file node: create the painted coverage result file and store it in a temporary folder (Use the paths stored in the file and package nodes to resolve those files)
   - return the tree of coverage results back to the controller
- start a file collection step on the controller that copies each painted file from the agent to the controller. Rather than using filenames use the hash code of (filename+path) to store the files (this is the way I avoid collisions in the warnings plugin)
- attach that result to the build as an action (on the controller)

**Comment by uhafner** [comment]
I finally found some time and started a new PR that will rewrite the source code painting: #265.

It is still work in progress but is much faster as the previous approach (and will not require a full filesystem scan of files).

**Comment by uhafner** [comment]
#265 is finally merged now. Here is an example for the new parameters:

https://github.com/jenkinsci/code-coverage-api-plugin/blob/2a443fbc3edb585f761829c05b3653543f1fc503/plugin/src/test/java/io/jenkins/plugins/coverage/model/CoveragePluginSourceITest.java#L156-L164

**Comment by notaturkey** [comment]
Awesome! Great work!
